handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,471 @@
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.DefaultEventLoopGroup;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.local.LocalAddress;
+import io.netty.channel.local.LocalChannel;
+import io.netty.channel.local.LocalServerChannel;
+import io.netty.handler.codec.DecoderException;
+import io.netty.handler.ssl.OpenSsl;
+import io.netty.handler.ssl.ReferenceCountedOpenSslContext;
+import io.netty.handler.ssl.ReferenceCountedOpenSslEngine;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import io.netty.util.ReferenceCountUtil;
+
+public class OcspTest {
+
+    /**
+     * The Server provides an OCSP staple and the Client accepts it.
+     */
+    @Test(timeout = 10000L)
+    public void testClientAcceptingOcspStaple() throws Exception {
+
+        assumeTrue(OpenSsl.isOcspSupported());
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        ChannelInboundHandlerAdapter serverHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                ctx.writeAndFlush(Unpooled.wrappedBuffer("Hello, World!".getBytes()));
+                ctx.fireChannelActive();
+            }
+        };
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                try {
+                    ReferenceCountUtil.release(msg);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ServerOcspContext serverOcspContext = newServerOcspContext();
+        TestClientOcspContext clientOcspContext = new TestClientOcspContext(true);
+
+        handshake(latch, serverOcspContext, clientOcspContext, serverHandler, clientHandler);
+
+        byte[] expected = serverOcspContext.staple();
+        byte[] actual = clientOcspContext.staple();
+
+        assertNotNull(actual);
+        assertNotSame(serverOcspContext.staple(), actual);
+        assertTrue(Arrays.equals(expected, actual));
+    }
+
+    /**
+     * The Server provides an OCSP staple and the Client rejects it.
+     */
+    @Test(timeout = 10000L)
+    public void testClientRejectingOcspStaple() throws Exception {
+
+        assumeTrue(OpenSsl.isOcspSupported());
+
+        final Throwable[] clientCauseRef = new Throwable[] { null };
```

Comment:
consider just using AtomicReference because you access this across different threads.